blaze-client/src/main/scala/org/http4s/client/blaze/ClientTimeoutStage.scala

```diff
@@ -0,0 +1,118 @@
+package org.http4s.client.blaze
+
+import java.nio.ByteBuffer
+import java.util.concurrent.TimeoutException
+import java.util.concurrent.atomic.AtomicReference
+
+import scala.annotation.tailrec
+import scala.concurrent.{Promise, Future}
+import scala.concurrent.duration.Duration
+import scala.util.{Failure, Success}
+
+import org.http4s.blaze.pipeline.MidStage
+import org.http4s.blaze.pipeline.Command.{EOF, Disconnect}
+import org.http4s.blaze.util.{ Cancellable, TickWheelExecutor }
+
+
+final private class ClientTimeoutStage(idleTimeout: Duration, requestTimeout: Duration, exec: TickWheelExecutor)
+  extends MidStage[ByteBuffer, ByteBuffer] 
+{ stage =>
+
+  private implicit val ec = org.http4s.blaze.util.Execution.directec
+  private var activeReqTimeout: Cancellable = Cancellable.noopCancel
+
+  override def name: String = s"ClientTimeoutStage: Idle: $idleTimeout, Request: $requestTimeout"
+
+  /////////// Private impl bits //////////////////////////////////////////
+  private val killswitch = new Runnable {
+    override def run(): Unit = {
+      logger.debug(s"Client stage is disconnecting due to timeout.")
+      // kill the active request: we've already timed out.
+      activeReqTimeout.cancel()
+
+      // check the idle timeout conditions
+      timeoutState.getAndSet(new TimeoutException(s"Client timeout.")) match {
+        case null => // noop
+        case c: Cancellable => c.cancel() // this should be the registration of us
+        case _: TimeoutException => // Interesting that we got here.
```

Comment:
Interesting enough to log?  This seems like the kind of statement one later regrets.